File: 3601-find-the-k-th-character-in-string-game-ii/3601-find-the-k-th-character-in-string-game-ii.go
package main

func helper(op []int, pos int64, k *int64) int {
	len := int64(1) << pos
	if len >= *k {
		return 0
	}
	res := helper(op, pos+1, k)
	if len < *k {
		*k -= len
		res = (res + op[pos]) % 26
	}
	return res
}

func kthCharacter(k int64, operations []int) byte {
	return byte('a' + helper(operations, 0, &k))
}

File: 3601-find-the-k-th-character-in-string-game-ii/3601-find-the-k-th-character-in-string-game-ii.cpp
class Solution {
    int helper(vector<int>& op, long long pos, long long& k) {
        long long len = 1ll<<pos;
        if(len >= k) return 0;
        int res = helper(op,pos + 1,k);
        if(len < k) {
            k -= len;
            res = (res + op[pos]) % 26;
        }
        return res;
    }
public:
    char kthCharacter(long long k, vector<int>& operations) {
        return 'a' + helper(operations,0,k);
    }
};